packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -3084,8 +3086,9 @@ namespace Tpetra {
         // We have to iterate through the row offsets anyway, so we
         // might as well check whether all rows' bounds are the same.
         bool allRowsReallySame = false;
+        auto k_rowPtrs_h = ::Tpetra::Details::getEntriesOnHost(this->k_rowPtrs_, 0, numRows);
```

Comment:
@kddevin or @brian-kelley -- I'm copying the whole array here because we know we're going to walk it anyway.  Are there any guarantees about when k_rowPtrs_ changes that allow us to save some performance by moving this copy out of this function to a place called less frequently?

Reply:
Yes, ``k_rowPtrs_`` should never change in between a fillComplete() and resumeFill(), since it can only change when entries are inserted or deleted from a row. ``lclGraph.row_map`` is const-valued to enforce this.

Reply:
so is there a problem with making a member view `k_rowPtrs_h_` that gets copied over during `fillComplete`?

What other functions might need to be called when the matrix is not fill-complete?  e.g. could we get rid of the `getEntryOnHost` and `getEntriesOnHost` calls throughout, or do some functions need to access the row pointers before fill-complete?

Reply:
@tasmith4 I think that's a good strategy - before fillComplete(), just use ``k_rowPtrs_h_`` so that ``getRowInfo()`` is very fast, and then copy it to device ``k_rowPtrs_`` during fillComplete when the local graph/matrix is being constructed. Then, ``k_rowPtrs_h_`` is always up-to-date whether fill complete or not.

Reply:
@brian-kelley so you're saying that all the performance-critical device things we need `k_rowPtrs_` to live on device for actually happen after fill-complete, so we don't need to be on device before then?

Reply:
@tasmith4 Right

Reply:
If you end up getting runtime errors from ``k_rowPtrs_`` being accessed on device before fillComplete, then you might have to make it a DualView or manually deep copy, but I don't know of anywhere that happens

Reply:
@brian-kelley do you know the difference between `fillComplete` and `expertStaticFillComplete`?

Reply:
I didn't know off the top of my head but:
3675     // Note: We don't need to do the following things which are normally done in fillComplete:
3676     // allocateIndices, globalAssemble, makeColMap, makeIndicesLocal, sortAndMergeAllIndices